Polymorphic Modules

LambdaAK opened the pull request:
Implemented Polymorphic modules.

LambdaAK commented on src/edu/cmu/cs/sasylf/ast/ContextJudgment.java:
```diff
@@ -31,8 +33,8 @@
  */
 public class ContextJudgment extends Judgment {
 
-	private final Judgment base;
```
I made these not final because we need to reassign to them when cloning.

LambdaAK commented on src/edu/cmu/cs/sasylf/ast/ModulePart.java:
```diff
@@ -54,6 +67,375 @@ public void typecheck(Context ctx) {
+			// We already verified above that resolution is an instance of Module
+			// Since Module is the only subclass of CompUnit, we know that resolution is an instance of CompUnit
+			// TODO: Add more checking here for casting safety
+			CompUnit functor = (CompUnit) resolution;
```
For the reasons specified in the comment, is it okay to just cast here?

LambdaAK commented on src/edu/cmu/cs/sasylf/ast/ModulePart.java:
```diff
@@ -54,6 +67,375 @@ public void typecheck(Context ctx) {
+
+				if (argResolution instanceof Syntax) {
+					Syntax argumentSyntax = (Syntax) argResolution;
+					SyntaxDeclaration parameterSyntax = (SyntaxDeclaration) parameterObject; // it will always be an instance of SyntaxDeclaration
```
Is it okay to cast here?

LambdaAK commented on src/edu/cmu/cs/sasylf/ast/QualName.java:
```diff
@@ -19,11 +22,23 @@
  * </ul>
  */
 public class QualName extends Node {
-	private final QualName source;
```
I made these not final because we need to reassign them when cloning

LambdaAK commented on src/org/sasylf/Activator.java:
```diff
@@ -76,11 +76,11 @@ public static Activator getDefault() {
 	 * @author VonC @ stackoverflow
 	 */
 	public Image getImage(String path) {
-		Image image = Activator.getDefault().getImageRegistry().get(path);
```
I don't think this file should be changed in this PR.

boyland replied:
That's right.  DOn't change it.

LambdaAK commented on src/org/sasylf/IDEProof.java:
```diff
@@ -216,7 +216,7 @@ public void dispose() {
 	 * @return proof object or null
 	 */
 	public static IDEProof getProof(IAdaptable res) {
-		return proofs.get(IProjectStorage.Adapter.adapt(res));
```
I don't think this file should be changed in this PR.

boyland replied:
That's right.  DOn't change it.

LambdaAK replied:
Fixed

LambdaAK commented on src/edu/cmu/cs/sasylf/ast/CompUnit.java:
```diff
@@ -224,11 +238,15 @@ public void collectQualNames(Consumer<QualName> consumer) {
 	public Object getDeclaration(Context ctx, String name) {
-		if (cacheVersion != ctx.version()) {
+		// TODO: We need to figure out how to handle the cache properly
+		if (true || cacheVersion != ctx.version()) {
```
Caching isn't done frequently enough, so I wrote `true ||` here to not use the cached data. I am still working on a way around this.

LambdaAK replied:
> Yes, currently, there are regression tests that fail if we get rid of `true ||`. I'm still trying to figure why this is the case, but, for now, having `true ||` eliminates the bug.

LambdaAK replied:
Fixed

boyland commented on src/edu/cmu/cs/sasylf/util/Span.java:
Ditto: is copy/substitute reasonable here?

boyland replied:
What case is this?  Span's as spans should be immutable.

LambdaAK replied:
In `Case.java`, there is a field `private final Span span`. When `Case.substitute` is called, `span` needs to be updated. Same applies for `Case.copy`.

boyland commented on src/org/sasylf/ProofChecker.java:
Don't change this.

boyland commented on src/org/sasylf/editors/MarkerResolutionGenerator.java:
Don;t change this.

boyland commented on src/org/sasylf/editors/ProofContentAssistProcessor.java:
Don't change this.

boyland commented on src/org/sasylf/editors/ProofEditor.java:
DOn't change this.

boyland commented on src/org/sasylf/editors/SASyLFAutoIndentStrategy.java:
Don't change this.

boyland commented on src/org/sasylf/editors/SASyLFColorProvider.java:
Don't change this.

boyland commented on src/org/sasylf/editors/SASyLFCorrectIndentStrategy.java:
Don't change this.

boyland commented on src/org/sasylf/editors/SASyLFIndentStrategy.java:
Don't change this.

boyland commented on src/org/sasylf/handlers/QuickFixHandler.java:
DOn't change this.

boyland commented on src/org/sasylf/project/ProjectProperties.java:
Don't change this.

boyland commented on src/org/sasylf/project/ProofBuilder.java:
Don't change this.

boyland commented on src/org/sasylf/refactor/MoveProofFile.java:
Don't change this.

boyland commented on src/org/sasylf/refactor/RenameProofModule.java:
Don't change this.

LambdaAK replied:
Fixed

boyland commented on src/org/sasylf/refactor/RenameTheoremWizard.java:
Don't change this.

LambdaAK replied:
Fixed

boyland commented on src/org/sasylf/util/QuickFixPage.java:
Don't change this.

LambdaAK replied:
Fixed

boyland commented on src/org/sasylf/util/TrackDirtyRegions.java:
Don't change this.

boyland commented on src/org/sasylf/views/ProofOutline.java:
Don't change this.

LambdaAK replied:
Fixed

boyland reviewed (changes requested):
Lots of good tests.  Directory structure umclear.
Some code problems -- see comments.
Eclipse plugin code should not be changed (it seems you had a refactoring get out of control -- VSC refactoring seems dumber than that in Eclipse?)

LambdaAK commented on run_tests.py:
```diff
@@ -0,0 +1,31 @@
+if len(sys.argv) > 1:
+  tests = [sys.argv[1]]
+
+regression_tests = [str(i) for i in range(58, 74 + 1)]
```
`run_tests.py` is an old testing file, so I will delete it.

LambdaAK commented on slf/aaaa.slf:
This was an old suite of tests I was using, but I have moved them over to `./regression`.

LambdaAK commented on src/edu/cmu/cs/sasylf/Proof.java:
Fixed

LambdaAK commented on src/edu/cmu/cs/sasylf/SubstitutionData.java:
```diff
@@ -0,0 +1,237 @@
+
+  }
+
+  private final IdentitySet<Object> set; // the set of objects that have been substituted for
```
Should they be private with `get` methods instead? Some of them need to be accessed externally.

LambdaAK commented on src/edu/cmu/cs/sasylf/CopyData.java:
```diff
@@ -0,0 +1,61 @@
+import java.util.IdentityHashMap;
+import java.util.Map;
+
+/** CloneData stores a mapping between objects and their clones, and has
```
Fixed

LambdaAK commented on src/edu/cmu/cs/sasylf/ast/Clause.java:
```diff
@@ -55,19 +58,31 @@ public void add(Element e) {
 
 	protected List<Element> elements = new ArrayList<Element>();
 
-	@Override
```
Having `implements Cloneable` in the declaration of `Clause` is redundant, because `Clause` implements `Element`, `Element` extends `Node`, `Node` implements `Span`, and `Span` implements `Cloneable`.

The reason we want to override `clone()` is to specialize the return type. If we don't override `clone()` in `Clause`, then the `clone()` that is nearest to `Clause` is the one inherited from `Object`, which returns `Object`. By overriding `clone()` in `Clause` we can cast the `Object` returned from `super.clone()` prior to returning it, which prevents us from needing to cast it outside of where `Clause.clone()` is called.

LambdaAK commented on src/edu/cmu/cs/sasylf/ast/CompUnit.java:
```diff
@@ -65,11 +74,13 @@ public boolean isAbstract() {
-	private List<Part> params = new ArrayList<Part>();
-	private List<Part> parts = new ArrayList<Part>();
-	private int parseReports;
+	public List<Part> getParams() {
```
I agree that returning a mutable list is suspicious. `getParams()` and `getParts()` are required because in `ModulePart`, we need to access the parameters and body of a functor when it is applied.

We should discuss how we can work around this.

LambdaAK replied:
`getParams()` and `getParts()` are no longer used anywhere, which is good for abstraction.

LambdaAK commented on src/edu/cmu/cs/sasylf/ast/RuleLike.java:
```diff
@@ -214,6 +215,9 @@ public Application checkApplication(Context ctx, List<? extends Fact> inputs, Fa
 			if (input instanceof DerivationPlaceholder) continue; // don't check
 			Element actual = input.getElement();
 			if (formal.getType().typeTerm() != actual.getType().typeTerm()) {
+				System.out.println("Not equal");
```
Fixed

LambdaAK commented on src/edu/cmu/cs/sasylf/ast/SyntaxDeclaration.java:
```diff
@@ -50,7 +54,7 @@ public SyntaxDeclaration(Location loc, NonTerminal nt, List<Clause> l) {
 		alternates = new TreeSet<String>();
 		alternates.add(Util.stripId(nonTerminal.getSymbol()));
-	}
+	}  
```
Fixed

LambdaAK commented on src/edu/cmu/cs/sasylf/ast/RenameSyntaxDeclaration.java:
```diff
@@ -57,7 +59,9 @@ public Variable getVariable() {
 
 	@Override
 	public void updateContext(Context ctx) {
+		// There's a bug somewhere between here
```
Fixed

LambdaAK commented on src/edu/cmu/cs/sasylf/ast/Syntax.java:
```diff
@@ -72,4 +74,24 @@ public void computeSubordination() {}
 	 */
 	public void checkSubordination() {}
+
+	/**
```
Agreed!

LambdaAK commented on src/edu/cmu/cs/sasylf/ast/SyntaxDeclaration.java:
```diff
@@ -408,11 +414,6 @@ public ClauseDef getContextClause() {
 		return context;
 	}
 
-	@Override
```
It should not have been removed. Fixed.

LambdaAK commented on src/edu/cmu/cs/sasylf/ast/TermPrinter.java:
Fixed.

LambdaAK commented on src/edu/cmu/cs/sasylf/ast/Terminal.java:
```diff
@@ -3,16 +3,20 @@
 import edu.cmu.cs.sasylf.util.Span;
 
 public class Terminal extends Element implements ElemType {
+	private Span sp;
```
I agree, it doesn't do anything. I removed it.

LambdaAK commented on .project:
```diff
@@ -31,4 +31,15 @@
 		<nature>org.eclipse.pde.PluginNature</nature>
 		<nature>org.eclipse.pde.api.tools.apiAnalysisNature</nature>
 	</natures>
+	<filteredResources>
```
Fixed

LambdaAK commented on src/edu/cmu/cs/sasylf/ast/DerivationByAssumption.java:
```diff
@@ -10,4 +11,13 @@ public DerivationByAssumption(String n, Location l, Clause c) {
+		return clone;
+	}
+
 }
```
`DerivationByAssumption` extends `Derivation`, and `Derivation` has `substitute()` implemented already. The reason we don't override `substitute()` in `DerivationByAssumption` is that `DerivationByAssumption` doesn't have any additional fields, therefore, the `substitute()` method in `DerivationByAssumption` would be the same as `Derivation`.

LambdaAK commented on src/edu/cmu/cs/sasylf/ast/DerivationByIHRule.java:
```diff
@@ -50,31 +51,37 @@ protected boolean acceptPlaceholder() {
 		try {
 			callSub = subject.unify(pattern);
-		} catch (UnificationIncomplete e) {
+		} catch (UnificationIncomplete e) {;
```
Fixed

LambdaAK commented on src/edu/cmu/cs/sasylf/ast/DerivationByIHRule.java:
```diff
@@ -50,31 +51,37 @@ protected boolean acceptPlaceholder() {
 					"(was checking " + subject + " instance of " + pattern + ",\n got exception " + e);      
 			return; // tell Java we're gone.
 		} catch (UnificationFailed e1) {
+			System.out.println("Failed");
```
Fixed

LambdaAK commented on src/edu/cmu/cs/sasylf/ast/DerivationByIHRule.java:
```diff
@@ -100,11 +107,18 @@ public void typecheck(Context ctx) {
 				learnAboutErrors.avoid(ctx.inputVars);
 				Term explanationTerm = learnAboutErrors.getSubstituted(concVar);
-				explanationString = tp.toString(tp.asClause(explanationTerm));
+				
```
This was originally a way to get around a bug that I had, but that bug has been resolved.

LambdaAK commented on src/edu/cmu/cs/sasylf/ast/Judgment.java:
```diff
@@ -16,10 +19,17 @@
+	private List<Rule> rules;
+	private Clause form;
+	private String name;
+	public NonTerminal assume;
```
Fixed

LambdaAK commented on src/edu/cmu/cs/sasylf/ast/JudgmentPart.java:
```diff
@@ -114,4 +118,49 @@ public void collectQualNames(Consumer<QualName> consumer) {
+		return sb.toString();
+	}
+
+	public List<Judgment> getJudgments() {
```
After the big refactoring of the `ModulePart` typechecking algorithm, we actually don't need `getJudgments()`. I think it still makes sense to have `getJudgments()` in the class so I will leave it, but modifying it such that it returns a shallow copy of the `judgments` field.

LambdaAK commented on src/edu/cmu/cs/sasylf/ast/ModulePart.java:
```diff
@@ -49,10 +97,379 @@ public void typecheck(Context ctx) {
+			// We already verified above that resolution is an instance of Module
+			// Since Module is the only subclass of CompUnit, we know that resolution is an instance of CompUnit
+			CompUnit functor = (CompUnit) resolution;
+
```
Done :)

LambdaAK commented on src/edu/cmu/cs/sasylf/ast/NonTerminal.java:
```diff
@@ -69,7 +101,7 @@ public void setType(SyntaxDeclaration t) {
 	}
 
-	private String symbol;
+	public String symbol;
```
There are places in the code where we need to clone a `NonTerminal`, but add filler characters to `symbol`. I fixed this by adding a method `cloneWithFillerCharacters` that does this. After that, I was able to make `symbol` private.

LambdaAK commented on src/edu/cmu/cs/sasylf/ast/QualName.java:
```diff
@@ -19,11 +23,53 @@
+		resolution = null;
+	}
+
+	public void setName(String newName) {
```
`setName()` was being used for substitution in `ModulePart`. I have refactored the code so that `QualName` substitution occurs in `QualName.substitute()`, which allowed me to get rid of `setName()` and make `name` private.

LambdaAK commented on src/edu/cmu/cs/sasylf/ast/RenameSyntaxDeclaration.java:
```diff
@@ -18,8 +20,8 @@
 public class RenameSyntaxDeclaration extends SyntaxDeclaration {
-	private final QualName source;
-	private SyntaxDeclaration original;
+	public QualName source;
```
Fixed

LambdaAK commented on src/edu/cmu/cs/sasylf/ast/SyntaxDeclaration.java:
```diff
@@ -63,7 +67,7 @@ public SyntaxDeclaration(Location loc, NonTerminal nt) {
 	public boolean isAbstract() { return isAbstract; }
 
-	private List<Clause> elements;
+	public List<Clause> elements; // productions
```
Fixed

LambdaAK commented on src/edu/cmu/cs/sasylf/ast/SyntaxDeclaration.java:
```diff
@@ -186,8 +191,8 @@ public void typecheck(Context ctx) {
 				elements.set(i, cd);
-				ctx.setProduction(cd.getConstructorName(),cd);
-				ctx.parseMap.put(cd.getElemTypes(), cd);
+				ctx.setProduction(cd.getConstructorName(),cd);Context.updateVersion();
```
Removed

LambdaAK commented on src/edu/cmu/cs/sasylf/term/Abstraction.java:
Yes, there should be copy/substitute because, if we copy any AST, we want it to be incompatible with the original one.

boyland replied:
But that should be done by making nbew LF, not by doing copy/substitute on LF.

LambdaAK replied:
Yes, I agree with you. The LF stuff needs to change, but it would be ideal if we could reassign rather than mutate. I will work on fixing this.

LambdaAK replied:
I'm pretty sure that LF term substitution only requires knowing `String from` and `String to`, where `from` is the name of whatever we're replacing, and `to` is the name of whatever we're inserting into the AST.

I'm thinking about making an abstract method
```java
public abstract Term substitute(String from, String);
```
in `Term`, and overriding it in `Abstraction`, `Application`, `Atom`, `BoundVar`, `Constant`, and `FreeVar`.

LambdaAK replied:
I have implemented this.

LambdaAK commented on src/edu/cmu/cs/sasylf/term/Application.java:
```diff
@@ -40,8 +45,8 @@ public Application(Atom f, Term a) {
 		this(f, Arrays.asList(new Term[] { a }));
 	}
 
-	private final Atom function;
```
Fixed

LambdaAK commented on src/edu/cmu/cs/sasylf/ast/Clause.java:
```diff
@@ -237,22 +252,58 @@ protected List<Element> withoutTerminals() {
 	 * @param orig original element
 	 * @param repl new element
 	 */
 	protected static void checkMatch(Element orig, Element repl) {
```
No, I removed it.

LambdaAK commented on src/edu/cmu/cs/sasylf/util/ErrorHandler.java:
```diff
@@ -183,4 +194,226 @@ public static Location lexicalErrorAsLocation(String file, String error) {
 			return new Location(file,0,0);
 		}
 	}
+
```
They are declared in the `Error` enum. The reason I added methods in `ErrorHandler` is because that allows for more specific feedback to the user. Typically, error messages have the form `<error_type, specific_info>`. Each method corresponds to a particular `error_type`, and the arguments are used to construct `specific_info`.

Using these methods allows us to abstract away the logic of crafting the error message when throwing it.

LambdaAK commented on src/edu/cmu/cs/sasylf/util/Location.java:
I agree that substitution doesn't make sense for `Location`. However, I think we need `copy()` when `DefaultSpan` and `TokenSpan` are copied.

LambdaAK commented on src/edu/cmu/cs/sasylf/ast/NonTerminal.java:
```diff
@@ -19,9 +21,20 @@
+	 * Substitutable is set to true by default, and should be set to false after the NonTerminal is substituted.
+	 * This is to prevent the NonTerminal from being replaced if there is a QualName module parameter with the same symbol.
+	 */
+	private boolean substitutable;
```
I have a few tests in `./regression` that test this, and they are all passing, so I think it works.

The idea is that once something is substituted for, we shouldn't need to substitute for it again.

For example, suppose we have the following module

```
module functor requires

abstract syntax a
abstract syntax b

provides

syntax pair ::= {a, b}
```

we can *roughly* model this mathematically as $\lambda a . \lambda  b . (a, b)$

and we use it as follows

```
module main provides

terminals T F S

syntax b ::= T | F
syntax n ::= 0 | S n

module instance = functor[b, n]
```

note what happens after the first substitution. When the `b` syntax is substituted for `a` in the body of the functor, the `b` syntax is captured by the `b` argument. So the functor after one application is $\lambda b . (b', b)$, where the `'` denotes that `b'` cannot be substituted for. We won't need to ever replace it with another syntax, So, I think setting `substitutable = false` won't cause any problems.

boyland commented on src/edu/cmu/cs/sasylf/CopyData.java:
Why in the sasylf package and not in sasylf.util ?

LambdaAK replied:
I moved it to `sasylf.util`.

boyland commented on src/edu/cmu/cs/sasylf/SubstitutionData.java:
ditto: why in top-level package?

LambdaAK replied:
I moved `SubstitutionData` to `ast`.

boyland commented on src/edu/cmu/cs/sasylf/ast/DerivationByIHRule.java:
```diff
@@ -50,31 +51,37 @@ protected boolean acceptPlaceholder() {
-		Util.debug("line: ", this.getLocation().getLine());
-
 		RuleLike ruleLike = getRule(ctx);
+
```
The code still has a comment about a bug.

boyland commented on src/edu/cmu/cs/sasylf/ast/Relaxation.java:
DOe sthis file still need to be changed?

LambdaAK replied:
No, Fixed.

boyland commented on tests/eight/demo.slf:
I thought tests/*** were out of date.

LambdaAK replied:
Yes, `test/***` is out of date. I have removed it.

boyland reviewed (changes requested):
Very close.
I think that the edu.cmu.cs.sasylf package (top-level) was for Facade objects, so I think CopyData and SubstitutionData should be in ....util.  But maybe you and Jonathan decided to put them in the top-level package?

Also: are the tests/ tests meant to be there?

And one or two other very minor things.

LambdaAK commented:
Hello John and Jonathan,
I have addressed all of these comments. For SubstitutionData, I think Jonathan and I agreed to put it into ast, since SubstitutionData operates on the AST. But, I think it could also go into util.
Let me know if you have any other feedback!
Best,
-Alex
    On Monday, August 5, 2024 at 11:53:50 AM EDT, John Boyland ***@***.***> wrote:  
 
 


@boyland requested changes on this pull request.

Very close.
I think that the edu.cmu.cs.sasylf package (top-level) was for Facade objects, so I think CopyData and SubstitutionData should be in ....util. But maybe you and Jonathan decided to put them in the top-level package?

Also: are the tests/ tests meant to be there?

And one or two other very minor things.

On src/edu/cmu/cs/sasylf/CopyData.java:

Why in the sasylf package and not in sasylf.util ?

On src/edu/cmu/cs/sasylf/SubstitutionData.java:

ditto: why in top-level package?

On src/edu/cmu/cs/sasylf/ast/Relaxation.java:

DOe sthis file still need to be changed?

On tests/eight/demo.slf:

I thought tests/*** were out of date.

In src/edu/cmu/cs/sasylf/ast/DerivationByIHRule.java:
>  		RuleLike ruleLike = getRule(ctx);
+

The code still has a comment about a bug.

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you authored the thread.Message ID: ***@***.***>

boyland commented:
Oh, yes.  I remember.  "ast" makes sense for Substitution Data.

Best regards,
John
________________________________
From: Alex Kozik ***@***.***>
Sent: Monday, August 5, 2024 6:03 PM
To: boyland/sasylf ***@***.***>
Cc: John T Boyland ***@***.***>; Mention ***@***.***>
Subject: Re: [boyland/sasylf] Polymorphic Modules (PR #128)


Hello John and Jonathan,
I have addressed all of these comments. For SubstitutionData, I think Jonathan and I agreed to put it into ast, since SubstitutionData operates on the AST. But, I think it could also go into util.
Let me know if you have any other feedback!
Best,
-Alex
On Monday, August 5, 2024 at 11:53:50 AM EDT, John Boyland ***@***.***> wrote:




@boyland requested changes on this pull request.

Very close.
I think that the edu.cmu.cs.sasylf package (top-level) was for Facade objects, so I think CopyData and SubstitutionData should be in ....util. But maybe you and Jonathan decided to put them in the top-level package?

Also: are the tests/ tests meant to be there?

And one or two other very minor things.

On src/edu/cmu/cs/sasylf/CopyData.java:

Why in the sasylf package and not in sasylf.util ?

On src/edu/cmu/cs/sasylf/SubstitutionData.java:

ditto: why in top-level package?

On src/edu/cmu/cs/sasylf/ast/Relaxation.java:

DOe sthis file still need to be changed?

On tests/eight/demo.slf:

I thought tests/*** were out of date.

In src/edu/cmu/cs/sasylf/ast/DerivationByIHRule.java:
> RuleLike ruleLike = getRule(ctx);
+

The code still has a comment about a bug.

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you authored the thread.Message ID: ***@***.***>

—
Reply to this email directly, view it on GitHub<https://github.com/boyland/sasylf/pull/128#issuecomment-2270063766>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAJDNU5DWDLMM4I4BCVGMVLZQAAFRAVCNFSM6AAAAABK7IV3MOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDENZQGA3DGNZWGY>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

boyland reviewed (approved):
ALl my comments were addressed.
